### PR TITLE
Refactor lower-level `EventSourcedRepo` functionality into `EventIndex` and `EventStore`

### PR DESCRIPTION
**Pull request by sygao**
- Refactor `src` write and load into `EventStore`, accessed through `EventSourcedRepo._src`
- Refactor OTU ID keyed event index and event index cache into `EventIndex`, accessed through `EventSourcedRepo._event_index`
- `EventIndexCache` now accessed solely through `EventSourcedRepo._cache`

### CODE REVIEW

File: virtool_cli/ref/repo.py [hunk 1]
```diff
@@ -596,14 +533,114 @@ def _load_otu_events_from_cache_and_update(self, otu_id: uuid.UUID) -> list[int]
...
+                continue
+        event_ids.sort()
+
+        return event_ids
+
+    def iter_events(self, reverse: bool = False):
+        for path in sorted(self.path.glob("*.json"), reverse=reverse):
+            if path.stem == "meta":
+                continue
+            yield _read_event_at_path(path)
```
Comment (igboyes) [inline virtool_cli/ref/repo.py:595]:
```suggestion
            if path.stem != "meta":
                yield _read_event_at_path(path)
```

File: virtool_cli/ref/repo.py [hunk 2]
```diff
@@ -596,14 +533,114 @@ def _load_otu_events_from_cache_and_update(self, otu_id: uuid.UUID) -> list[int]
...
+
+    def iter_events_from_index(self, start: int = 1) -> Generator[Event, None, None]:
```
Comment (igboyes) [inline virtool_cli/ref/repo.py:597]: Do you really need this and `iter_events`?

File: virtool_cli/ref/repo.py [hunk 3]
```diff
@@ -596,14 +533,114 @@ def _load_otu_events_from_cache_and_update(self, otu_id: uuid.UUID) -> list[int]
...
             return otu_event_list
 
         return []
 
+    def get_otu_metadata(self, event_ids: list[int]) -> dict | None:
+        """Retrieves OTU metadata from a list of event IDs"""
+        if not event_ids:
+            return None
+        event_ids.sort()
+        first_event_id = event_ids[0]
```
Comment (igboyes) [inline virtool_cli/ref/repo.py:547]: Avoid mutating your function parameter.
```suggestion
        first_event_id = sorted(event_ids)[0]
```

File: virtool_cli/ref/repo.py [hunk 4]
```diff
@@ -125,10 +119,14 @@ def new(cls, data_type: DataType, name: str, path: Path, organism: str):
 
         return EventSourcedRepo(path)
 
+    @property
+    def last_id(self):
```
Comment (igboyes) [inline virtool_cli/ref/repo.py:123]: Docstring.

**Review by igboyes — CHANGES_REQUESTED** [review]
I think you should fix these minor things and move towards merge. I think any further changes should come in an additional PR.

File: virtool_cli/ref/repo.py (hunk 2 above)
Comment (sygao) [reply virtool_cli/ref/repo.py:597]: They're used in multiple places and were there in ESRepo before, so removing them would take some time... it's mostly `.generate_after_start()` that is hanging out without a use case at the moment.

**Pull request closed without merging**